### ISSUE TITLE
NCCL GPU affinity (nvidia-smi topo -m) on VM with fail PIX,PHB, PXB GDRDMA. but good performance on BM.

### ISSUE THREAD

**Issue by dobiup** [NVIDIA/nccl#1464]
Hi, I need your help seriously.

First, NVIDIA HW/SW component for intra-node and inter-node communication are working well as point of compatibility, enabled.
NVSwitch (NVlink, Fabric Manager), GPU (GPU Driver, CUDA, NCCL), HCA(OFED, nv_peer_mem or nvidia_peermem) and so on.

NCCL 2.23.4 for CUDA 12.2 on ProxMox Hypervisor.
Ubuntu 22.04.5 GPU Driver for DC 535.183.06 Mellanox OFED 23.10-3.2.2.
nv_peer_mem 535.183.06 (Assembled in GPU driver) : installed manually too.
 
InfiniBand RDMA performance is also good (Not GPU Direct RDMA), the same as BareMetal environment according to IB_send, IB_receive BW, Latency results.
The problem is the efficient GPU Direct RDMA (PIX, PHX, PXB) does not work in NCCL due to the intra-node topology recognized by the VM, as explained below.
As you know, GPURDMA has performance effects only when the HCA of GPU memory is transmitted through the PCIe switch.
 
NCCL log (As you can see, GPU Direct RDMA was enabled, but then disabled due to topology differences.)
  Enabled
![Image](https://github.com/user-attachments/assets/7f96cee7-3ea1-43b0-be50-b2b21370f65c)
Disabled
![Image](https://github.com/user-attachments/assets/0eb33e89-495b-42a9-8ee9-96384b7089e6)

 
So, we checked by expanding GPU Direct RDMA to system memory copy level (via CPU – which does not have a significant performance advantage)
NCCL_NET_GDR_LEVEL=SYS :
NCCL log
![Image](https://github.com/user-attachments/assets/e1783574-45a3-4e51-ab1f-b544bc2a81cc)
GPU Direct RDMA via system memory copy
![Image](https://github.com/user-attachments/assets/8da8fcac-d766-46b5-8d8f-e237800fd454)

Currently, what we most want is the same intro-node topology (nvidia-smi topo -m) as BareMetal,
But, with the help of NVIDIA, we want to make GPU Direct RDMA works at the PIX, PXB, and PHB levels first.
 
The key is how to configure a VM environment (GPU Affinity with CPU Affinity) that allows topology (PIX, PHB, PXB) in GPU Direct RDMA.

VM intra topology
![Image](https://github.com/user-attachments/assets/7da2accc-5aa5-43c9-aede-f6f50fc00c4d)

BM intra topology
![Image](https://github.com/user-attachments/assets/b0a3fa7b-4326-45e2-ba13-96122c0d04f5)

mpirun \
    -np 16 \
    -N 8 \
    --bind-to socket \
    --hostfile /home/singtel/nccl-tests/hosts-2 \
    -x NCCL_IB_CUDA_SUPPORT=1 \
    -x CUDA_VISIBLE_DEVICES=0,1,2,3,4,5,6,7 \
    -x LD_LIBRARY_PATH \
    -x NCCL_UCX_TLS=rc_x,cuda_copy \
    -x NCCL_UCX_RNDV_THRESH=0 \
    -x UCX_MEMTYPE_CACHE=n \
    -x NCCL_COLLNET_ENABLE=0 \
    -x NCCL_PLUGIN_P2P=ucx \
    -x NCCL_DEBUG=info \
    -x NCCL_DEBUG_SUBSYS=NET \
    -x NCCL_IB_HCA=mlx5 \
    /home/singtel/nccl-tests/build/all_reduce_perf -b 128 -e 16G -f 2 -g 1 -n 50 -w 100 -p 0 -z 0 -t 1 -c 1

![Image](https://github.com/user-attachments/assets/8ad3d190-8dee-4b45-a2d1-b99a7efd5616)
![Image](https://github.com/user-attachments/assets/dc485160-7b1e-451e-97ef-0ff3d7835b24)

**Comment by sjeaugey** [comment]
So, are you asking how to inject a topology inside NCCL?

Would that comment help:
https://github.com/NVIDIA/nccl-tests/issues/86#issuecomment-848939744

**Comment by dobiup** [comment]
Thank you for update.

Yes, Exactly what I want. Where can I get the NCCL_TOPO_FILE ?
Only VM provider (ProxMox or KVM) can provide this?

**Comment by dobiup** [comment]
Hi, SJ

Where can I find the NCCL topology xml file? I couldn't find it from below?


NCCL_TOPO_FILE
(since 2.6)
Path to an XML file to load before detecting the topology. By default, NCCL will load /var/run/nvidia-topologyd/virtualTopology.xml if present.


    // Try default XML topology location
    NCCLCHECKGOTO(ncclTopoGetXmlFromFile("/var/run/nvidia-topologyd/virtualTopology.xml", xml, 0), ret, fail);

**Comment by sjeaugey** [comment]
If your cloud provider doesn't provide one (or if you are launching your VM yourself), you'd need to write it, based on the physical topology.

**Comment by tchaton** [comment]
Hey @dobiup. Did you ever managed to make it work. I have generated the TOPO file but unfortunately, I am seeing the same performance degradation in the VM

**Comment by tchaton** [comment]
Hey @sjeaugey, I am curious if you know if it is possible to get good NCCL performance in VM for multi node training

**Comment by sjeaugey** [comment]
Yes it's certainly possible, but the question is quite broad. There are a number of things you'd need to do aside from the topology injection, including ACS/ATS configuration for GPUDirect transfers to not be funneled to the CPU root complex, and ensuring the network and GPU work at full speed inside the VM.

I'm not an expert at that unfortunately, this is VM/HW configuration skills I don't have.

**Comment by tchaton** [comment]
Hey @sjeaugey That's really interesting to know. I am using Cloud Hypervisor: https://github.com/cloud-hypervisor/cloud-hypervisor/issues/8141

I had found those docs: https://docs.nvidia.com/ai-enterprise/planning-resource/optimizing-vm-configuration-ai-inference/latest/appendix.html. 

I did the topology injection, performed the ACS/ATS configuration on the host.

The results with ib_write_bw were almost in parity with baremetal.

```
Per-NIC RDMA Bandwidth (ib_write_bw)


| Test | Path | Bandwidth | % of NDR ceiling | Status |
|------|------|----------:|----------------:|:------:|
| Cross-node ALL 8 NICs | A.mlx5_[0-7] ↔ B.mlx5_[0-7] | 3,160 Gb/s | 99.7%/NIC | 🏆 fabric bisection
| Cross-node fabric, 4 QP / 1MB | Node A mlx5_0 ↔ Node B mlx5_0 | 395.08 Gb/s | 99.7% | 🏆 |
| NIC-pair, 4 QP / 1MB (1 node) | mlx5_0 → mlx5_1 | 395.07 Gb/s | 99.7% | 🏆 |
| NIC-pair, single QP (1 node) | mlx5_X → mlx5_(X+1) | 368 Gb/s | 93% | ✅ |
| Loopback (NUMA-pinned, 65KB) | mlx5_0 → mlx5_0 | 290 Gb/s | 73% | ✅ |
```

NCCL All-Reduce (8 GB message)

```
| Configuration | GPUs | Avg Bus BW | Wrong values | Path |
|---|:---:|---:|:---:|---|
| Single node (Host 1) | 8 | 474.11 GB/s | 0 | NVLink/NVSwitch | ✅ 
| Single node (Host 2) | 8 | 474.02 GB/s | 0 | NVLink/NVSwitch | ✅ 
| 2-node (16 GPUs) | 16 | 154.925 GB/s | — | NVLink + IB fabric | ❌
| 2-node (16 GPUs) Baremetal + ACS | 16 | 186 GB/s  | NVLink + IB fabric | ❌
| 2-node (16 GPUs) Baremetal - ACS | 16 | 456.109 GB/s | NVLink + IB fabric | ✅
```

But yeah, no luck on making Multi Node NCCL close to baremetal speed within the VM. Would you know anyone with that expertize ?

**Comment by sjeaugey** [comment]
It would have been good to open a new issue; it's not clear your issue is related to the original question, plus this bug has been closed for some time.

Now it would looks like ATS is not functional on your system: when ACS is enabled, all traffic is redirected to the CPU root complex causing a bottleneck when all NICs communicate concurrently. You may not see the perf issue on ib_write_bw given ACS may be able to sustain a single flow at full speed. With ACS+ATS, in theory, most of the traffic no longer needs to go to the RC and performance would be close to baremetal.
